src/charm.py

```diff
@@ -210,21 +280,95 @@ def _configure(self, event: EventBase) -> None:
             return
+        if not self._is_ca_cert_active():
+            logger.info("Renewing CA certificate")
+            self._renew_root_certificate()
+            return
         self._send_ca_cert()
         self._process_outstanding_certificate_requests()
 
+    def _renew_root_certificate(self):
+        """Generate a new active root CA certificate.
+
+        If there is an active CA certificate that is about to expire, generate a new one.
+        If there is no active CA certificate, generate a new one.
+        """
+        if not self.unit.is_leader():
+            return
+        self._move_active_ca_cert_to_expiring()
+        self._generate_root_certificate()
+
+    def _move_active_ca_cert_to_expiring(self):
+        """Make current active CA certificate expiring.
+
+        CA certificate is moved to the secret holding expiring CA certificates.
+        The validity of the expiring CA can't be shorter than
+            the validity of the issued certificates.
+        """
+        try:
+            current_active_ca_cert_secret = self.model.get_secret(
+                label=CA_CERTIFICATES_SECRET_LABEL
+            )
+            current_active_ca_cert_secret_content = current_active_ca_cert_secret.get_content(
+                refresh=True
+            )
+        except SecretNotFoundError:
+            logger.warning("No active CA certificate found to move to expiring")
+            return
+        self._set_juju_secret(
+            label=EXPIRING_CA_CERTIFICATES_SECRET_LABEL,
+            content=current_active_ca_cert_secret_content,
+            expire=self._config_certificate_validity,
+        )
+
+    def _set_juju_secret(self, label: str, content: dict[str, str], expire: timedelta) -> None:
+        """Create or update a juju secret."""
+        try:
+            secret = self.model.get_secret(label=label)
+            secret.set_content(content=content)
+            secret.set_info(expire=expire)
+            secret.get_content(refresh=True)
+        except SecretNotFoundError:
+            self.app.add_secret(content=content, label=label, expire=expire)
+
+    # TODO
     def _root_certificate_matches_config(self) -> bool:
         """Return whether the stored root certificate matches with the config."""
         if not self._config_ca_common_name:
             raise ValueError("CA common name should not be empty")
         ca_certificate_secret = self.model.get_secret(label=CA_CERTIFICATES_SECRET_LABEL)
         ca_certificate_secret_content = ca_certificate_secret.get_content(refresh=True)
         ca = ca_certificate_secret_content["ca-certificate"]
-        return self._config_ca_common_name == Certificate.from_string(ca).common_name
+        certificate = Certificate.from_string(ca)
+        certificate_validity = (
+            certificate.expiry_time - certificate.validity_start_time
+            if certificate.validity_start_time and certificate.expiry_time
+            else timedelta(days=0)
```

Comment:
I'm not sure why we have this value in the else statement.

Reply:
This should never happen, but if the certificate has no validity values set we consider it's validity to be for 0 days, triggering a renewal flow as this function will return False.
This was necessary because these values are optional in the lib in the `Certificate` object.